### PR TITLE
refactor: update Android bridge

### PR DESCRIPTION
**Pull request by Mercy811**
- Update Android bridge file at `android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt`
- Add Android native side tests at `android/src/test/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPluginTest.kt`
  - Simulate method call from Flutter 
  - Verify corresponding Android APIs are called
 

To run Android native unit tests `cd example/android && ./gradlew testDebugUnitTest`
Unfortunately I cannot make Android unit test CI running. Will just keep the tests to local for now. 
image

### CODE REVIEW

File: android/build.gradle [hunk 1]
```diff
@@ -2,14 +2,14 @@ group 'com.amplitude.amplitude_flutter'
 version '1.0-SNAPSHOT'
 
 buildscript {
-    ext.kotlin_version = '1.7.10'
+    ext.kotlin_version = '1.5.20'
```
Comment (izaaz) [inline android/build.gradle:5]: Are we going to an earlier version of kotlin?

File: lib/amplitude.dart [hunk 2]
```diff
@@ -147,7 +147,7 @@ class Amplitude extends _Amplitude {
       event.mergeEventOptions(options);
     }
 
-    return await _channel.invokeMethod("track", jsonEncode(event.toMap()));
```
Comment (izaaz) [inline lib/amplitude.dart:150]: Why does this change? Was the old version calling the wrong method or did the interface change in the new Android SDK?

**Review by izaaz — APPROVED** [review]

File: android/build.gradle [hunk 3]
```diff
@@ -44,6 +44,20 @@ android {
 }
 dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
-    implementation "com.amplitude:android-sdk:2.39.8"
-    implementation "com.squareup.okhttp3:okhttp:4.2.2"
```
Comment (yuhao900914) [inline android/build.gradle:48]: Are we going to implement the migration logic? Or should that be done in other tickets?

File: android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt [hunk 4]
```diff
@@ -263,236 +139,157 @@ class AmplitudeFlutterPlugin : FlutterPlugin, MethodCallHandler {
...
 
-                    // UNSET
-                    "\$unset" -> {
-                        identify.unset(key)
-                    }
+    internal fun getEvent(json: JSONObject): BaseEvent {
+        val plan = json.getJSONObject("plan")
+        val ingestionMetadata = json.getJSONObject("ingestion_metadata")
+        val event = BaseEvent()
+        event.eventType = json.getString("event_type")
```
Comment (yuhao900914) [inline android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt:203]: Do we need to check if  the input has event_type or not?

File: android/build.gradle (hunk 3 above)
Comment (Mercy811) [reply android/build.gradle:48]: What is the migration logic for?

File: lib/amplitude.dart (hunk 2 above)
Comment (Mercy811) [reply lib/amplitude.dart:150]: This is to fix the error of my last commit. Note this PR is compared with branch `4.x` not `main`. The interface `setGroup()` remains unchanged.

File: android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt (hunk 4 above)
Comment (Mercy811) [reply android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt:203]: `eventType` is a required for `BaseEvent` constructor. So skip checking whether it's existing or not.

File: android/build.gradle (hunk 1 above)
Comment (Mercy811) [reply android/build.gradle:5]: I think I lowered it for some compatibility issues. I don't see any harms to customers to lower the version. Do you any concern? I'm open to discussion and can try to see if I can revert it if it's necessary.

File: android/build.gradle (hunk 3 above)
Comment (yuhao900914) [reply android/build.gradle:48]: The migration from the older flutter to the latest flutter.

Comment (Mercy811) [reply android/build.gradle:48]: Data migration in Kotlin SDK is turned on by default https://www.docs.developers.amplitude.com/data/sdks/android-kotlin/migration/#data-migration. Could you elaborate on the what we need to migrate?

Comment (yuhao900914) [reply android/build.gradle:48]: NVM. The flutter is a bridge of the latest ios sdk and latest android sdk. The migration is over there.

**Review by yuhao900914 — APPROVED** [review]

File: android/build.gradle [hunk 5]
```diff
@@ -25,7 +25,7 @@ apply plugin: 'com.android.library'
 apply plugin: 'kotlin-android'
 
 android {
-    compileSdkVersion 28
+    compileSdkVersion 31
```
Comment (justin-fiedler) [inline android/build.gradle]: Since we are updating any reason not to use latest? e.g. 34

File: android/build.gradle [hunk 6]
```diff
@@ -44,6 +44,20 @@ android {
 }
 dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
-    implementation "com.amplitude:android-sdk:2.39.8"
-    implementation "com.squareup.okhttp3:okhttp:4.2.2"
+    implementation "org.jetbrains.kotlin:kotlin-reflect:1.6.10"
+
+    implementation 'com.amplitude:analytics-android:1.+'
```
Comment (justin-fiedler) [inline android/build.gradle]: nit: This sometimes gives warnings I have switched to Maven style ranges instead

https://github.com/amplitude/session-replay-android/blob/main/plugin-session-replay/build.gradle.kts#L42

File: android/build.gradle [hunk 7]
```diff
@@ -2,14 +2,14 @@ group 'com.amplitude.amplitude_flutter'
...
-    ext.kotlin_version = '1.7.10'
+    ext.kotlin_version = '1.5.20'
     repositories {
         google()
         mavenCentral()
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.6.4'
+        classpath 'com.android.tools.build:gradle:7.0.4'
```
Comment (justin-fiedler) [inline android/build.gradle]: If we are updating can we use latest? e.g. gradle 8.2.x

File: android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt [hunk 8]
```diff
@@ -1,260 +1,136 @@
...
+                val defaultTracking = json.getJSONObject("defaultTracking")
+                amplitude = Amplitude(
+                    Configuration(
+                        apiKey = json.getString("apiKey"),
+                        context = ctxt!!,
+                        flushQueueSize = json.getInt("flushQueueSize"),
+                        flushIntervalMillis = json.getInt("flushIntervalMillis"),
+                        instanceName = json.getString("instanceName"),
+                        optOut = json.getBoolean("optOut"),
+                        minIdLength = if (json.optInt("minIdLength") == 0) defaultMinIdLength else json.optInt(
```
Comment (justin-fiedler) [inline android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt]:
```suggestion
                        minIdLength = if (json.optInt("minIdLength") < 1) defaultMinIdLength else json.optInt(
```

File: android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt [hunk 9]
```diff
@@ -263,236 +139,157 @@ class AmplitudeFlutterPlugin : FlutterPlugin, MethodCallHandler {
...
-                    }
-                }
+    internal fun JSONObject.toMutableMap(): MutableMap<String, Any?> {
+        val map = mutableMapOf<String, Any?>()
+        val keys = keys()
+        while (keys.hasNext()) {
+            val key = keys.next()
+            var value = get(key)
+            if (value is JSONObject) {
+                value = value.toMap()
```
Comment (justin-fiedler) [inline android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt:275]: Should this be mutable too? e.g. `value.toMutableMap()`

File: example/android/build.gradle [hunk 10]
```diff
@@ -1,12 +1,12 @@
...
-    ext.kotlin_version = '1.6.0'
+    ext.kotlin_version = '1.6.21'
     repositories {
         google()
         mavenCentral()
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:4.0.1'
+        classpath 'com.android.tools.build:gradle:7.4.2'
```
Comment (justin-fiedler) [inline example/android/build.gradle]: Can we update this to match the gradle version used by the SDK?

**Review by justin-fiedler — APPROVED** [review]

File: android/build.gradle (hunk 5 above)
Comment (Mercy811) [reply android/build.gradle]: Updated to 34 . Where do you normal check the latest version? I searched "compileSdkVersion latest version" but didn't get an answer to be 34.

File: android/build.gradle (hunk 7 above)
Comment (Mercy811) [reply android/build.gradle]: `7.0.4` was the suggested version by Android Studio somehow. Updated to `classpath 'com.android.tools.build:gradle:8.2.2'`.

File: android/build.gradle (hunk 6 above)
Comment (Mercy811) [reply android/build.gradle]: Update to `implementation 'com.amplitude:analytics-android:[1.0,2.0)'`

File: android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt (hunk 9 above)
Comment (Mercy811) [reply android/src/main/kotlin/com/amplitude/amplitude_flutter/AmplitudeFlutterPlugin.kt:275]: This function will be removed in my next PR.

File: example/android/build.gradle (hunk 10 above)
Comment (Mercy811) [reply example/android/build.gradle]: Updated to 8.2.2.

File: android/build.gradle (hunk 5 above)
Comment (justin-fiedler) [reply android/build.gradle]: Thanks. I knew 34 was latest from the SR repo. I don't know the best place to check but this seems to be up to date
https://developer.android.com/tools/releases/platforms

File: example/android/gradle/wrapper/gradle-wrapper.properties [hunk 11]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.7.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-bin.zip
```
Comment (justin-fiedler) [inline example/android/gradle/wrapper/gradle-wrapper.properties]:
```suggestion
distributionUrl=https\://services.gradle.org/distributions/gradle-8.2.2-bin.zip
```

File: example/android/app/build.gradle [hunk 12]
```diff
@@ -21,12 +21,13 @@ if (flutterVersionName == null) {
...
 }
 
+apply plugin: 'idea'
 apply plugin: 'com.android.application'
 apply plugin: 'kotlin-android'
 apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
 
 android {
-    compileSdkVersion 28
+    compileSdkVersion 33
```
Comment (justin-fiedler) [inline example/android/app/build.gradle]: Make this 34 too?

File: example/android/gradle/wrapper/gradle-wrapper.properties (hunk 11 above)
Comment (Mercy811) [reply example/android/gradle/wrapper/gradle-wrapper.properties]: Updated in https://github.com/amplitude/Amplitude-Flutter/pull/171/commits/1a3d02c98c31522b0dab94ea4cf6ce6d1e8f75f1.

File: example/android/app/build.gradle (hunk 12 above)
Comment (Mercy811) [reply example/android/app/build.gradle]: Updated in https://github.com/amplitude/Amplitude-Flutter/pull/171/commits/1a3d02c98c31522b0dab94ea4cf6ce6d1e8f75f1.

**Pull request merged**